frame/democracy/src/lib.rs

```diff
@@ -1049,8 +1085,107 @@ pub mod pallet {
+					let _ = T::ExternalOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SuperMajorityAgainst => {
+					let _ = T::ExternalDefaultOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SimpleMajority => {
+					let _ = T::ExternalMajorityOrigin::ensure_origin(origin)?;
+				},
+			};
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T>::PreimageNotExist);
+			T::Preimages::request(&hash);
+			MetadataOf::<T>::insert(MetadataOwner::External, hash);
+			Self::deposit_event(Event::<T>::MetadataSet { owner: MetadataOwner::External, hash });
+			Ok(())
+		}
+
+		/// Set the metadata of the public proposal.
+		///
+		/// - `origin`: Must be `Signed`, and the creator of the proposal.
+		/// - `index`: The index of the proposal.
+		/// - `hash`: The preimage hash of an on-chain stored preimage.
+		#[pallet::weight(T::WeightInfo::set_proposal_metadata())]
+		pub fn set_proposal_metadata(
+			origin: OriginFor<T>,
+			index: PropIndex,
+			hash: PreimageHash,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (_, _, proposer) = Self::proposal(index)?;
+			ensure!(proposer == who, Error::<T>::NoPermission);
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T>::PreimageNotExist);
+			T::Preimages::request(&hash);
+			let owner = MetadataOwner::Proposal(index);
+			MetadataOf::<T>::insert(owner.clone(), hash);
+			Self::deposit_event(Event::<T>::MetadataSet { owner, hash });
+			Ok(())
+		}
+
+		/// Clear the external proposal metadata.
+		///
+		/// - `origin`: Must be an external origin and correspond to the proposal vote threshold,
+		///   `ExternalOrigin` for `SuperMajorityApprove`, `ExternalDefaultOrigin` for
+		///   `SuperMajorityAgainst` and `ExternalMajorityOrigin` for `SimpleMajority`.
+		#[pallet::weight(T::WeightInfo::clear_external_metadata())]
+		pub fn clear_external_metadata(origin: OriginFor<T>) -> DispatchResult {
+			let (_, threshold) = <NextExternal<T>>::get().ok_or(Error::<T>::NoProposal)?;
+			match threshold {
+				VoteThreshold::SuperMajorityApprove => {
+					let _ = T::ExternalOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SuperMajorityAgainst => {
+					let _ = T::ExternalDefaultOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SimpleMajority => {
+					let _ = T::ExternalMajorityOrigin::ensure_origin(origin)?;
+				},
+			};
+			Self::clear_metadata(MetadataOwner::External);
+			Ok(())
+		}
+
+		/// Clear the public proposal metadata.
+		///
+		/// - `origin`: Must be `Signed`, and the creator of the proposal.
+		/// - `index`: The index of the proposal.
+		#[pallet::weight(T::WeightInfo::clear_proposal_metadata())]
+		pub fn clear_proposal_metadata(origin: OriginFor<T>, index: PropIndex) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let (_, _, proposer) = Self::proposal(index)?;
+			ensure!(proposer == who, Error::<T>::NoPermission);
+			Self::clear_metadata(MetadataOwner::Proposal(index));
+			Ok(())
+		}
```

Comment:
Anyone should be able to clear the proposal metadata when the proposal is no longer in the `publicProps`. So there should be a check that if the proposal index isnt in the list, that we can immediately remove the metadata without authorization check.

Reply:
same note about `set_metadata` accepting a `none`

Reply:
@shawntabrizi can you explain why anyone can clear metadata of ongoing referendum?

How about anyone can clear metadata of not ongoing referendum?
Metadata of ongoing referendum is immutable.